Skip Soft delete using move API

dheerajsngh opened the pull request:
Implementation

1. Implementation of move for JSON and gRPC client
2. Added config flag to switch to the new implementation
3. Added unit tests for the respective implementation
4. Added integration test

Testing :

1. Unit tests running
2. Tested locally for JSON and gRPC for flat bucket and folder renames,with and without the flags

animesh-g commented:
/gcbrun

animesh-g commented on gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientTest.java:
```diff
@@ -451,6 +491,178 @@ public void copyObjects_differentBuckets_succeeds() throws Exception {
+
+        gcs.move(moveMap);
+        fail("Expected UnsupportedOperationException was not thrown.");
+      } catch (UnsupportedOperationException e) {
```
We can't we use assertThrows() here instead of manually catching exceptions

dheerajsngh replied:
Updated the tests. PTAL

animesh-g commented on gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientTest.java:
For this whole file, format comments like sentences. i.e. comments should be valid english sentences ending with periods.

dheerajsngh replied:
Updated. 
PTAL

dheerajsngh replied:
Updated comments to add periods

animesh-g commented on gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientTest.java:
```diff
@@ -451,6 +491,178 @@ public void copyObjects_differentBuckets_succeeds() throws Exception {
+
+        gcs.move(moveMap);
+      fail("Expected FileNotFoundException was not thrown.");
+    } catch (FileNotFoundException e) {
```
Same as below. We should prefer using assertThrows.

dheerajsngh replied:
Done

animesh-g commented on gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientTest.java:
```diff
@@ -70,9 +72,32 @@ public class GoogleCloudStorageClientTest {
   private static final StorageResourceId TEST_RESOURCE_ID =
       new StorageResourceId(TEST_BUCKET_NAME, TEST_OBJECT_NAME);
 
+  private static final StorageResourceId TEST_RESOURCE_ID_2 =
```
nit: Why don't we follow same pattern of unit tests of Copy API i.e. appending `move` prefix in destination for the sake of consistency?

dheerajsngh replied:
Updated the tests. PTAL

animesh-g commented on gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:
```diff
@@ -222,6 +222,39 @@ public MethodOutcome renameFileIntoRootOutcome() {
+      googleHadoopFileSystem.delete(new Path(initUri));
+    }
+  }
+
```
ABove one only tests for JSON path. We should also add tests for gRPC(i.e. enabling STORAGE_CLIENT).

dheerajsngh replied:
Enabled for gRPC as well. Thanks for pointing it out.

animesh-g reviewed (approved):
Approving this. Please make sure tests are passed before merging.

dheerajsngh commented on gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientTest.java:
```diff
@@ -617,34 +611,42 @@ public void move_emptyMap_doesNothing() throws Exception {
+          assertThrows(UnsupportedOperationException.class,
+              () ->
+                  gcs.move(moveMap));
+      assertThat(thrown).hasMessageThat()
```
Looking at the documentation, it seems that the third argument is not used for verifying the error message but is instead when the [assert fails](https://screenshot.googleplex.com/BebDyAiFe6RYNzP).

animesh-g commented:
gcbrun/

animesh-g commented:
/gcbrun

animesh-g commented:
/gcbrun

animesh-g commented:
/gcbrun

dheerajsngh commented:
FYI : Removed the integration test in the last [commit](https://github.com/GoogleCloudDataproc/hadoop-connectors/pull/1322/commits/8307e6c57553b424d9379b3fe472a8e41988d424) since the GCP project where the tests run is not whitelisted for move API. Once the project is whitelisted, we will add the integration tests back.

singhravidutt commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:
```diff
@@ -1102,6 +1141,61 @@ public void copy(Map<StorageResourceId, StorageResourceId> sourceToDestinationOb
+   * about expected behavior.
+   */
+  @Override
+  public void move(Map<StorageResourceId, StorageResourceId> sourceToDestinationObjectsMap)
```
So, what if some moves were successful and some are not. Client can't just retry the request as it will get FileNotFound for all where move was successful in the first request.

dheerajsngh replied:
Won't this be similar to the cases where copy has succeeded for all objects but delete succeeds for some objects and fails for some of the objects? We had discussed this [[thread](https://chat.google.com/room/AAAAsZJYxUI/ux_LLjkhLZQ/ux_LLjkhLZQ?cls=10)/]

dheerajsngh replied:
Updated the implementation to not throw exception for file Not found cases.

singhravidutt commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/testing/InMemoryGoogleCloudStorage.java:
```diff
@@ -326,6 +326,12 @@ public void deleteFolders(List<FolderInfo> folders) throws IOException {
+  @Override
+  public synchronized void move(
+      Map<StorageResourceId, StorageResourceId> sourceToDestinationObjectsMap) throws IOException {
+    throw new IOException("Not implemented");
```
Please do add the basic implementation for testing.

dheerajsngh replied:
Sure, will add the test.

singhravidutt commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java:
```diff
@@ -479,6 +481,105 @@ public void copy(
+    }
+
+    if (storageOptions.getEncryptionKey() != null) {
+      moveRequestBuilder.setSourceOptions(
```
why this is required? Can this be passed in as headers? If so we do have interceptor for it.

dheerajsngh replied:
In [grpc](https://github.com/search?q=repo%3AGoogleCloudDataproc%2Fhadoop-connectors%20getEncryptionKey&type=code) we are mostly passing it along with the request and not in the header.

singhravidutt reviewed (commented):
Mostly looks fine. Might wanna rethink about the scenario where some move operation succeed and some fail.

dheerajsngh commented on gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java:
```diff
@@ -556,6 +556,12 @@ public class GoogleHadoopFileSystemConfiguration {
           AsyncWriteChannelOptions.DEFAULT.getPartFileNamePrefix());
 
+  /** Configuration key for enabling move operation in gcs instead of copy+delete. */
+  public static final HadoopConfigurationProperty<Boolean> GCS_OPERATION_MOVE_ENABLE =
```
Yes we plan to update the `Configuration.md` in May, once `move` API is GA.

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:
```diff
@@ -993,6 +993,45 @@ public static void validateCopyArguments(
+      // Avoid move across buckets.
+      if (!srcBucketName.equals(dstBucketName)) {
+        throw new UnsupportedOperationException(
+            "This operation is not supported across two different buckets.");
```
For copy operation as well,we are using the same message,so will keep it same for consistency for now. We can update for both in a follow up PR.

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:
```diff
@@ -993,6 +993,45 @@ public static void validateCopyArguments(
+        throw new UnsupportedOperationException(
+            "This operation is not supported across two different buckets.");
+      }
+      checkArgument(
```
This case will not happen, since for empty object names, src object names are appended before the current code is called. [code link/](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/195aaeed1a3bf255bcb75ea01b6dcd2d8f0f7851/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemImpl.java#L707)
[screenshot/](https://screenshot.googleplex.com/3LDt7hvPpY3EdRq)

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:
```diff
@@ -993,6 +993,45 @@ public static void validateCopyArguments(
+          !isNullOrEmpty(source.getObjectName()), "srcObjectName must not be null or empty");
+      checkArgument(
+          !isNullOrEmpty(destination.getObjectName()), "dstObjectName must not be null or empty");
+      if (srcBucketName.equals(dstBucketName)
```
We already check if dst and src are the same during the initial validation at the GoogleCloudStorageFileSystemImpl([cs/](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/195aaeed1a3bf255bcb75ea01b6dcd2d8f0f7851/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemImpl.java#L592)) layer before calling checking it here again.

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:
```diff
@@ -1102,6 +1141,61 @@ public void copy(Map<StorageResourceId, StorageResourceId> sourceToDestinationOb
+
+    validateMoveArguments(sourceToDestinationObjectsMap);
+
+    if (sourceToDestinationObjectsMap.isEmpty()) {
```
We are already doing it in validation. We are double checking it here again for early exit. Will remove the check here.

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:
```diff
@@ -1102,6 +1141,61 @@ public void copy(Map<StorageResourceId, StorageResourceId> sourceToDestinationOb
+   * about expected behavior.
+   */
+  @Override
+  public void move(Map<StorageResourceId, StorageResourceId> sourceToDestinationObjectsMap)
```
For the case with single src and single dst, we convert it to this function similar to how it is done for [copy](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/195aaeed1a3bf255bcb75ea01b6dcd2d8f0f7851/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemImpl.java#L617)

dheerajsngh commented:
Updated the implementation to include the following
1. Added Generation id check for source as well.
2. The implementation now handles the 'file not found' scenario without throwing an exception. This modification supports a retry mechanism, consistent with the approach used for deleteObjects.

will be updating the unit tests, in upcoming commits.

@singhravidutt , Please review the updated implementation.

singhravidutt commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemImpl.java:
```diff
@@ -764,6 +776,34 @@ private void renameDirectoryInternal(FileInfo srcInfo, URI dst) throws IOExcepti
+      // Finally, move marker items (if any) to mark rename operation success
+      moveInternal(srcToDstMarkerItemNames);
+
+      List<FileInfo> bucketToDelete = new ArrayList<>();
+      List<FileInfo> objectToDelete = new ArrayList<>();
+
+      if (srcInfo.getItemInfo().isBucket()) {
+        bucketToDelete.add(srcInfo);
+        deleteBucket(bucketToDelete);
+      } else {
+        // If src is a directory then srcItemInfos does not contain its own name,
+        // we delete item separately in the list.
+        objectToDelete.add(srcInfo);
+        deleteObjects(objectToDelete);
```
It seems we are deleting the srcInfo, if its an object we call deleteObject if it's a bucket we call deleteBucket. This whole logic of creating a list and then adding item to can be simplified.

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorage.java:
```diff
@@ -281,6 +281,17 @@ default void copy(Map<StorageResourceId, StorageResourceId> sourceToDestinationO
+   * Moves metadata of the given objects. Moving between two different buckets is not allowed.
+   *
+   * @param sourceToDestinationObjectsMap map of destination objects to be moved, keyed by source
+   * @throws java.io.FileNotFoundException if the source object or the destination bucket does not
```
Updated.

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java:
```diff
@@ -479,6 +481,116 @@ public void copy(
+      String srcObjectName,
+      long dstContentGeneration,
+      String dstObjectName) {
+    MoveBlobRequest.Builder moveRequestBuilder =
```
Done.

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemImpl.java:
```diff
@@ -614,14 +615,25 @@ public void rename(URI src, URI dst) throws IOException {
-                  srcInfo.getItemInfo().getObjectName(),
-                  srcInfo.getItemInfo().getContentGeneration())));
+      if (this.options.getCloudStorageOptions().isMoveOperationEnabled()
+          && srcResourceId.getBucketName().equals(dstResourceId.getBucketName())) {
```
Yes, i also had the same understanding, but while looking at the code we did not see any explicit check to stop cross bucket operations. 
While trying the operation using hadoop, we observed that move was not allowed but copy was allowed across bucket.
Hence we felt putting an explicit check seemed better.

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/ForwardingGoogleCloudStorage.java:
```diff
@@ -152,6 +152,13 @@ public void copy(Map<StorageResourceId, StorageResourceId> sourceToDestinationOb
     delegate.copy(sourceToDestinationObjectsMap);
   }
 
```
ACK. 
Will investiagte and update in a follow up PR.

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:
```diff
@@ -1211,6 +1306,64 @@ public void onFailure(GoogleJsonError jsonError, HttpHeaders responseHeaders) {
+            GoogleJsonResponseException cause =
+                createJsonResponseException(jsonError, responseHeaders);
+
+            if (errorExtractor.itemNotFound(cause)) {
```
ACK, as discussed, we will update in a follow up PR.

ShreyasSinha commented:
/gcbrun

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:
```diff
@@ -1211,6 +1306,71 @@ public void onFailure(GoogleJsonError jsonError, HttpHeaders responseHeaders) {
+                    ? createFileNotFoundException(bucketName, srcObjectName, cause)
+                    : new IOException(
+                        String.format(
+                            "Error moving '%s'",
```
Makes sense to include both. Updated
Thanks.

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/testing/InMemoryGoogleCloudStorage.java:
```diff
@@ -326,6 +326,75 @@ public void deleteFolders(List<FolderInfo> folders) throws IOException {
+        if (!srcInfo.exists()) {
+          // If the source is not found, add an error to the list and continue.
+          innerExceptions.add(
+              new IOException(String.format("Source object '%s' not found.", srcObject)));
```
Updated.

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/testing/InMemoryGoogleCloudStorage.java:
```diff
@@ -326,6 +326,75 @@ public void deleteFolders(List<FolderInfo> folders) throws IOException {
   }
 
+  @Override
+  public synchronized void move(
```
Simplified the logic a little bit to just do copy and delete not do any  extra checks.

dheerajsngh commented on gcsio/src/test/java/com/google/cloud/hadoop/gcsio/ForwardingGoogleCloudStorageTest.java:
```diff
@@ -219,6 +221,18 @@ public void testCopy() throws IOException {
+  @Test
+  public void testMove() throws IOException {
+    Map<StorageResourceId, StorageResourceId> sourceToDestinationObjectsMap =
+        new HashMap<>(TEST_STRINGS.size());
```
I think we can remove it.
Updated.

dheerajsngh commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java:
```diff
@@ -479,6 +481,96 @@ public void copy(
+   * about expected behavior.
+   */
+  @Override
+  public void move(Map<StorageResourceId, StorageResourceId> sourceToDestinationObjectsMap)
```
Agreed. I think there is a lot of code saharing which we can do between JSOn and gRPC implementation. 
I have created a task for this in our taskflow. Will take it in follow up PR.

singhravidutt reviewed (approved):
There are multiple followup after this adding them here so that we don't loose track of it
1. Adding integration tests as soon s we are able to run them in cloud build
2. Adding support of move API in PerformanceCache
3. Refactoring of batch logic for move and copy API across GoogleCloudStorage implementations.

ShreyasSinha commented:
/gcbrun
